kaldb/src/main/java/com/slack/kaldb/server/OpenSearchBulkIngestAPI.java

```diff
@@ -0,0 +1,237 @@
+      List<IndexRequest> indexRequests = OpenSearchRequest.parseBulkRequest(bulkRequest);
+      Map<String, List<Trace.Span>> docs =
+          OpenSearchRequest.convertIndexRequestToTraceFormat(indexRequests);
+      // our rate limiter doesn't have a way to acquire permits across multiple datasets
+      // so today as a limitation we reject any request that has documents against multiple indexes
+      // We think most indexing requests will be against 1 index
+      if (docs.keySet().size() > 1) {
+        BulkIngestResponse response =
+            new BulkIngestResponse(0, 0, "request must contain only 1 unique index");
+        return HttpResponse.ofJson(INTERNAL_SERVER_ERROR, response);
+      }
+
+      for (Map.Entry<String, List<Trace.Span>> indexDocs : docs.entrySet()) {
+        final String index = indexDocs.getKey();
+        if (!rateLimiterPredicate.test(index, indexDocs.getValue())) {
+          BulkIngestResponse response = new BulkIngestResponse(0, 0, "rate limit exceeded");
+          return HttpResponse.ofJson(TOO_MANY_REQUESTS, response);
+        }
+      }
+      BulkIngestResponse response = produceDocuments(docs);
+      return HttpResponse.ofJson(response);
+    } catch (Exception e) {
+      LOG.error("Request failed ", e);
+      BulkIngestResponse response = new BulkIngestResponse(0, 0, e.getMessage());
+      return HttpResponse.ofJson(INTERNAL_SERVER_ERROR, response);
+    }
+  }
+
+  public BulkIngestResponse produceDocuments(Map<String, List<Trace.Span>> indexDocs) {
+    int totalDocs = indexDocs.values().stream().mapToInt(List::size).sum();
+
+    String transactionId = String.join("_", indexDocs.keySet()) + "_" + System.currentTimeMillis();
+    // we cannot create a generic pool of producers because the kafka API expects the transaction ID
+    // to be a property while creating the producer object.
+    try (KafkaProducer<String, byte[]> kafkaProducer =
+        createKafkaTransactionProducer(transactionId)) {
+      for (Map.Entry<String, List<Trace.Span>> indexDoc : indexDocs.entrySet()) {
+        String index = indexDoc.getKey();
+        // call once per batch and use the same partition for better batching
+        int partition = getPartition(index);
+
+        // since there isn't a dataset provisioned for this service/index we will not index this set
+        // of docs
+        if (partition < 0) {
+          LOG.warn("index=" + index + " does not have a provisioned dataset associated with it");
+          continue;
+        }
+        try {
+          kafkaProducer.initTransactions();
+          kafkaProducer.beginTransaction();
+          for (Trace.Span doc : indexDoc.getValue()) {
+
+            ProducerRecord<String, byte[]> producerRecord =
+                new ProducerRecord<>(
+                    preprocessorConfig.getDownstreamTopic(), partition, index, doc.toByteArray());
+            kafkaProducer.send(producerRecord);
+          }
+          kafkaProducer.commitTransaction();
```

Comment:
Should this be part of the finally? What happens if an exception occurs and the transaction is not committed?

Reply:
good question. I forgot to add the close on finally . Updating PR now.

The commitTransaction can throw ProducerFencedException which we catch and cannot retry. However we don't retry anyways on the `KafkaException` catch block today